Leanplum/Leanplum.m

```diff
@@ -0,0 +1,152 @@
+//
+//  Leanplum.m
+//  Leanplum
+//
+//  Created by Hrishikesh Amravatkar on 5/9/19.
+//  Copyright © 2019 Leanplum. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import "Leanplum.h"
+#import "LPUserApi.h"
+#import "LPApiConstants.h"
+#import "LPUtils.h"
+
+@implementation Leanplum
+
++ (void)setApiHostName:(NSString *)hostName
+       withServletName:(NSString *)servletName
+              usingSsl:(BOOL)ssl
+{
+    if ([LPUtils isNullOrEmpty:hostName]) {
+        [self throwError:@"[Leanplum setApiHostName:withServletName:usingSsl:] Empty hostname "
+         @"parameter provided."];
+        return;
+    }
+    if ([LPUtils isNullOrEmpty:servletName]) {
+        [self throwError:@"[Leanplum setApiHostName:withServletName:usingSsl:] Empty servletName "
+         @"parameter provided."];
+        return;
+    }
+    
+    [LPApiConstants sharedState].apiHostName = hostName;
+    [LPApiConstants sharedState].apiServlet = servletName;
+    [LPApiConstants sharedState].apiSSL = ssl;
+}
+
++ (void)setUserAttributes:(NSDictionary *)attributes
+              withSuccess:(void (^)(void))success
+              withFailure:(void (^)(NSError *error))failure
+{
+    [self setUserId:@"" withUserAttributes:attributes withSuccess:success withFailure:failure];
+}
+
++ (void)setUserId:(NSString *)userId
+      withSuccess:(void (^)(void))success
+      withFailure:(void (^)(NSError *error))failure {
+    [self setUserId:userId withUserAttributes:@{} withSuccess:success withFailure:failure];
+}
+
++ (void)setUserId:(NSString *)userId withUserAttributes:(NSDictionary *)attributes
+      withSuccess:(void (^)(void))success
+      withFailure:(void (^)(NSError *error))failure
+{
+    attributes = [self validateAttributes:attributes named:@"userAttributes" allowLists:YES];
+    [self setUserIdInternal:userId withAttributes:attributes
+                withSuccess:(void (^)(void))success
+                withFailure:(void (^)(NSError *error))failure];
+}
+
++ (void)setUserIdInternal:(NSString *)userId
+           withAttributes:(NSDictionary *)attributes
+              withSuccess:(void (^)(void))success
+              withFailure:(void (^)(NSError *error))failure
+{
+    // Some clients are calling this method with NSNumber. Handle it gracefully.
+    id tempUserId = userId;
+    if ([tempUserId isKindOfClass:[NSNumber class]]) {
+        userId = [tempUserId stringValue];
+    }
+    
+    // Attributes can't be nil
+    if (!attributes) {
+        attributes = @{};
+    }
+    [LPUserApi setUsersAttributes:userId withUserAttributes:attributes success:^{
+        success();
+    } failure:^(NSError *error) {
+        failure(error);
+    }];
+}
+
++ (NSDictionary *)validateAttributes:(NSDictionary *)attributes named:(NSString *)argName
```

Comment:
can we put comment //copied from old SDK if we copy code exactly from old sdk? This way there are a lot fewer questions about certain things are done this way

Reply:
Thanks for the comment @gg4race . As per our meeting discussion. We will cherry pick and not reinvent wheel at some places like these . And lets review the code even thought its copied form somewhere else. No need to add comment as //copied etc.